pkg/endpoint/directory.go

```diff
@@ -48,27 +49,48 @@ func (e *Endpoint) backupDirectoryPath() string {
 
 // moveNewFilesTo copies all files, that do not exist in newDir, from oldDir.
 func moveNewFilesTo(oldDir, newDir string) error {
-	oldFiles, err := os.ReadDir(oldDir)
+	oldDirFile, err := os.Open(oldDir)
 	if err != nil {
 		return err
 	}
-	newFiles, err := os.ReadDir(newDir)
+	defer oldDirFile.Close()
+
+	oldFiles, err := oldDirFile.Readdirnames(-1)
+	if err != nil {
+		return err
+	}
+
+	newDirFile, err := os.Open(newDir)
+	if err != nil {
+		return err
+	}
+	defer newDirFile.Close()
+
+	newFiles, err := newDirFile.Readdirnames(-1)
 	if err != nil {
 		return err
 	}
 
+	newFilesHash := make(map[string]bool, len(newFiles))
+	for _, f := range newFiles {
+		newFilesHash[f] = true
+	}
+
+	wg := &sync.WaitGroup{}
+
 	for _, oldFile := range oldFiles {
-		exists := false
-		for _, newFile := range newFiles {
-			if oldFile.Name() == newFile.Name() {
-				exists = true
-				break
-			}
-		}
-		if !exists {
-			os.Rename(filepath.Join(oldDir, oldFile.Name()), filepath.Join(newDir, oldFile.Name()))
+		if ok := newFilesHash[oldFile]; !ok {
+			oldFile := oldFile
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				os.Rename(filepath.Join(oldDir, oldFile), filepath.Join(newDir, oldFile))
+			}()
```

Comment:
I'm not sure how many `oldFiles` there can be but this is spinning off as many goroutines as there are `oldFiles`. While the body of the goroutine is quite small--just renaming a file, this can put additional pressure onto the Go runtime if it is already managing many goroutines or if it is already under CPU pressure. We probably want to bound this number of goroutines to at least the number of CPUs, but even then this code is running for each endpoint that's regenerating, which can happen in parallel.

Reply:
I agree it would be nice to understand how many `oldFiles` there can be. I don't think the number of goroutines matters that much tbh, but we can limit concurrency here for sure - even something constant should be fine, like 5 or 10, as this is mostly io operation. Also, it would be nice to handle error from `os.Rename` as we haven't done that before...

Reply:
I pushed some changes to use an `errgroup.Group` to limit the number of goroutines to five and allow for us to bubble-up errors from `os.Rename`. Five goroutines is a small enough amount that it won't have a drastic impact on the runtime, and since these operations are IO bound we don't need to worry about these goroutines contributing to CPU pressure.

@christarazi helped me look into the number of `oldFiles` that we could expect and it should stay less than 10. There's four static files that are present here with an additional three that could be created if debug mode is enabled.

Hey @aanm, it looks like you originally authored this function five years ago in https://github.com/cilium/cilium/commit/f6c4385a43943c7be71bd8ca3a0582cab24db20b. Do you remember if there was a reason for not checking the error on `os.Rename`?

Reply:
@learnitall no specific reason.

Reply:
I didn't look to much into this PR but from what I can see the total number of go routines will depend on the number of endpoints on the node times ~4 (number of old files).

Reply:
I did a quick test to see what would happen if we kept the other changes in this PR, but run `os.Rename` in serial, and I don't see a difference. The sec/op stays about the same, and memory usage looks better since we skip creating the goroutines:

```
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/endpoint
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
                   │ MoveNewFilesTo1-before--.txt │   MoveNewFilesTo1-after.txt   │
                   │            sec/op            │   sec/op     vs base          │
MoveNewFilesTo1-20                    5.508µ ± 0%   5.514µ ± 1%  ~ (p=0.363 n=10)

                   │ MoveNewFilesTo1-before--.txt │     MoveNewFilesTo1-after.txt      │
                   │             B/op             │    B/op     vs base                │
MoveNewFilesTo1-20                     438.0 ± 0%   373.0 ± 0%  -14.84% (p=0.000 n=10)

                   │ MoveNewFilesTo1-before--.txt │     MoveNewFilesTo1-after.txt     │
                   │          allocs/op           │ allocs/op   vs base               │
MoveNewFilesTo1-20                     11.00 ± 0%   10.00 ± 0%  -9.09% (p=0.000 n=10)
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/endpoint
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
                   │ MoveNewFilesTo5-before--.txt │   MoveNewFilesTo5-after.txt   │
                   │            sec/op            │   sec/op     vs base          │
MoveNewFilesTo5-20                    6.376µ ± 1%   6.383µ ± 1%  ~ (p=0.897 n=10)

                   │ MoveNewFilesTo5-before--.txt │     MoveNewFilesTo5-after.txt     │
                   │             B/op             │    B/op     vs base               │
MoveNewFilesTo5-20                     705.0 ± 0%   641.0 ± 0%  -9.08% (p=0.000 n=10)

                   │ MoveNewFilesTo5-before--.txt │     MoveNewFilesTo5-after.txt     │
                   │          allocs/op           │ allocs/op   vs base               │
MoveNewFilesTo5-20                     20.00 ± 0%   19.00 ± 0%  -5.00% (p=0.000 n=10)
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/endpoint
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
                    │ MoveNewFilesTo10-before--.txt │  MoveNewFilesTo10-after.txt   │
                    │            sec/op             │   sec/op     vs base          │
MoveNewFilesTo10-20                     7.694µ ± 4%   7.644µ ± 1%  ~ (p=0.089 n=10)

                    │ MoveNewFilesTo10-before--.txt │     MoveNewFilesTo10-after.txt      │
                    │             B/op              │     B/op      vs base               │
MoveNewFilesTo10-20                    1.385Ki ± 0%   1.322Ki ± 0%  -4.51% (p=0.000 n=10)

                    │ MoveNewFilesTo10-before--.txt │    MoveNewFilesTo10-after.txt     │
                    │           allocs/op           │ allocs/op   vs base               │
MoveNewFilesTo10-20                      31.00 ± 0%   30.00 ± 0%  -3.23% (p=0.000 n=10)```

Reply:
Given the above, I'll go ahead and remove usage of goroutines to make this PR safer.